### DISCUSSION TITLE
Internal working/Cost function minimization or implementation of Mask thing while Registration

### DISCUSSION THREAD

**Discussion by anmolnamdev** [SuperElastix/elastix#866] [Q&A]
Hello everyone, 
I am currently using the elastix for the Non-rigid registration of 2 volumes of Head & Neck cancer dataset. I am also using the Fixed and Moving volumes Masks to skip registrations in some portions. 
Now the non-rigid registration without a mask is simple to understand and quite intuitive as the cost function which is being minimized is known.
But when I am using the masks, I want to know as to how internally it is working/implemented. **I am not able to figure out the internal procedure going on while non-rigid registration of 2 volumes along with the fixed and moving volumes Masks. 
Can someone help me in understanding the internals and theory behind them. It is really very urgent.**
Hope to get a response soon.

**Comment by ntatsisk** [comment]
Hi @anmolnamdev,
Have you taken a look at chapter 5.4 of the manual https://elastix.lumc.nl/doxygen/index.html? It gives some details on the masks.

As far as I understand, the masks allow to exlude specific parts of the image from being used to draw samples from. After the samples are drawn, the metric calculation procedure is the same as the case of no masks. You can find an older discussion related to that here: https://groups.google.com/g/elastix-imageregistration/c/wG9_jUeA_48

**Reply by anmolnamdev** [reply]
I read the chapter 5.4 from the manual. The issue is that I am not able to understand the end-to-end procedure of the registration being done when masks are involved. Can you explain me or point to any resource. That would be of lot of help. Please help.

**Reply by anmolnamdev** [reply]
One additional follow-up question to this is: If we exclude the masked region from the registration, then exactly how the continuity is maintained at the boundaries of the masks(by boundary, I mean, let us say I have a square image with small circular shape masked out, then how the continuity is maintained at the edges of this circle, so that it doesn't seem like this region has been pasted or explicitly added).
